core/include/jiminy/core/robot/AbstractConstraint.h

```diff
@@ -0,0 +1,90 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+///
+/// \brief      Generic interface for a kinematic constraint.
+///
+/// \details    This class enables the support of kinematics constraints on the system.
+///             The most common constraint is to say that a given frame is fixed: this
+///             is implemented in FixedFrameConstraint
+///
+/// \remarks    Each constraint applied to a system is downcasted as an instance of
+///             AbstractMotorBase and polymorphism is used to call the actual implementations.
+///
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+#ifndef JIMINY_ABSTRACT_CONSTRAINT_H
+#define JIMINY_ABSTRACT_CONSTRAINT_H
+
+#include <memory>
+
+#include "jiminy/core/Types.h"
+
+
+namespace jiminy
+{
+    class Model;
+
+    class AbstractConstraint: public std::enable_shared_from_this<AbstractConstraint>
+    {
+
+    public:
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief      Forbid the copy of the class
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        AbstractConstraint(AbstractConstraint const & abstractMotor) = delete;
+        AbstractConstraint & operator = (AbstractConstraint const & other) = delete;
+
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief      Constructor
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        AbstractConstraint();
+        virtual ~AbstractConstraint(void);
+
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        /// \brief      Initialize the constraint on the given model.
+        ///
+        /// \note       This function is called internally when adding a constraint to a robot:
+        ///             there is no need to call it manually.
+        /// \param[in] model    Model on which to apply the constraint.
+        ///////////////////////////////////////////////////////////////////////////////////////////////
+        virtual hresult_t initialize(Model *model);
```

Comment:
Initialized is in fact used as an "attach" method. I provide no "detach" method yet but it may be a good thing to have one if only for consistency.

I'm currently exposing ```get_jacobian``` and ```get_drift``` in python for debugging - but the main usage I see is to only create the constraint in python, and add it to a robot - then leave the object be.

Reply:
The difference for me is that there can be several types of constraints, unlike contact points or external forces. This makes it similar to a motor / sensor. Also, constraint constructors have different parameters (and some have several like the wheel constraint we propose). I don't see an easy way to handle that if the object is created by the Robot class.

But I agree with you that it would make more sense to manage the object by Robot. In a way this is already what is being done using shared_pointers.  The typical use case I see is for the user to create the object in python, then give it to the robot (in fact, I may hide get_jacobian / get_drift from the python bindings such that the only accessible method is the constructor itself). Then if the user releases his pointer, the only remaining instance is in the Robot - such that if you remove the constraint from the robot the object is destroyed. The only time when it's not the case is if the user explicitly kept his instance of shared pointer for whatever reason - and I don't know then if it's justified to destroy the object for him (potentially he wants to keep memory of the constraint he created for attaching it to another robot...).

Of course, the harmful side effect is if you try to attach the same constraints to two robots.

So I propose to have a behavior similar to the sensors with an ```uninitialize``` function (similar to detach) for the constraint. On model removal the constraint is first uninitialized - and you cannot add an already initialized constraint to a robot.

Reply:
After reading other comments: I'm all the more in favor of handling constraints like motors and sensors, as objects to be plugged in on a robot (not a model), with an initialized / uninitialize mechanism.

Reply:
> But I agree with you that it would make more sense to manage the object by Robot.

You've convinced me.

I think keeping `get_jacobian / get_drift ` as read-only attribute in Python is nice, at least for debug and quick-and-dirty prototyping.

> Of course, the harmful side effect is if you try to attach the same constraints to two robots.
> So I propose to have a behavior similar to the sensors with an uninitialize function (similar to detach) for the constraint. On model removal the constraint is first uninitialized - and you cannot add an already initialized constraint to a robot.

I don't like the idea of having an `uninitialize` method. I would rather use a mutex (as I did to lock the Robot once the simulation has started). If the mutex is locked, then it is impossible to call `initialize`. Removing the constraint would release the lock, and thereby enable to use `initialize` again.

Reply:
Just add a get method, to go with add and remove. To be consistent with sensor/motor API.